src/utils/utf16.rs

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2017 fd developers
+// Licensed under the Apache License, Version 2.0
+// <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0>
+// or the MIT license <LICENSE-MIT or http://opensource.org/licenses/MIT>,
+// at your option. All files in the project carrying such
+// notice may not be copied, modified, or distributed except
+// according to those terms.
+
+use std::iter::{Peekable, Rev};
+
+#[derive(Debug)]
+pub struct ReverseUtf16Iterator<'a, I>
+where
+    I: 'a + DoubleEndedIterator<Item = &'a u16>,
+{
+    it: Peekable<Rev<I>>,
+    nx: Option<u16>,
+}
+
+impl<'a, I> ReverseUtf16Iterator<'a, I>
+where
+    I: DoubleEndedIterator<Item = &'a u16>,
+{
+    pub fn new(it: I) -> Self {
+        ReverseUtf16Iterator {
+            it: it.rev().peekable(),
+            nx: None,
+        }
+    }
+}
+
+impl<'a, I> Iterator for ReverseUtf16Iterator<'a, I>
+where
+    I: DoubleEndedIterator<Item = &'a u16>,
+{
+    type Item = u16;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(next) = self.nx {
+            self.nx = None;
+            Some(next)
+        } else {
+            if let Some(chr) = self.it.next() {
+                if let Some(next_chr) = self.it.next() {
+                    if (next_chr >= &&0xD800u16) && (next_chr <= &&0xDFFFu16) {
```

Comment:
Could you add a short comment here (I have no idea what is going on here :smile:)?

Reply:
This is an implementation of a reverse iterator over a valid UTF-16 sequence. Since UTF-16 encodes some characters on 2 words, it's not possible to simply reverse the sequence, as some word pairs must be kept in the same order. I added a small comment about that.